[gradle] (flaky?) Jackson-related error when building image

<!-- Thanks for filing an issue! Please use this template to provide more context for the issue . -->

**Description of the issue**: Got a random jackson exception when building with `jib`.

Exception when building image. Retrying the same build worked.

`404 not found` seems to be the root cause, which I'm assuming is flakiness when accessing `gcr.io`. I want to just get the log out here, but there might not be anything to do here. It's why I'm keeping this issue lightweight.

```
Step #3 - "curio-generated-build-all": Caused by: com.google.cloud.tools.jib.plugins.common.BuildStepsExecutionException: Build image failed
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.plugins.common.BuildStepsRunner.build(BuildStepsRunner.java:269)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.gradle.BuildImageTask.buildImage(BuildImageTask.java:119)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Step #3 - "curio-generated-build-all": at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Step #3 - "curio-generated-build-all": at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
Step #3 - "curio-generated-build-all": ... 50 more
Step #3 - "curio-generated-build-all": Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `com.google.cloud.tools.jib.registry.json.ErrorResponseTemplate` (although at least one Creator exists): no int/Int-argument constructor/factory method to deserialize from Number value (404)
Step #3 - "curio-generated-build-all": at [Source: (String)"404 page not found
Step #3 - "curio-generated-build-all": "; line: 1, column: 1]
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1343)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1032)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromInt(ValueInstantiator.java:262)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromInt(StdValueInstantiator.java:356)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromNumber(BeanDeserializerBase.java:1324)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:173)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:161)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4013)
Step #3 - "curio-generated-build-all": at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3004)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.json.JsonTemplateMapper.readJson(JsonTemplateMapper.java:101)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.registry.RegistryEndpointCaller.call(RegistryEndpointCaller.java:248)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.registry.RegistryEndpointCaller.callWithAllowInsecureRegistryHandling(RegistryEndpointCaller.java:152)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.registry.RegistryEndpointCaller.call(RegistryEndpointCaller.java:142)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.registry.RegistryClient.callRegistryEndpoint(RegistryClient.java:339)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.registry.RegistryClient.pullManifest(RegistryClient.java:200)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.registry.RegistryClient.pullManifest(RegistryClient.java:208)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.pullBaseImage(PullBaseImageStep.java:197)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call(PullBaseImageStep.java:168)
Step #3 - "curio-generated-build-all": at com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.call(PullBaseImageStep.java:56)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
Step #3 - "curio-generated-build-all": at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
```

**Expected behavior**:

Retries so the build doesn't fail? Or possibly provides a clearer exception message

**Steps to reproduce**: 

Running a jib build

**Environment**: Linux, Gradle 5.0-rc-4, Java 11

**`jib-gradle-plugin` Configuration**: 

A relatively standard configuration that pushes to Google Container Registry

**Log output**: <!-- If applicable, provide relevant log output -->

**Additional Information**: <!-- Any additional information that may be helpful -->


<!-- Thanks for contributing! -->

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java
@@ -263,17 +263,7 @@ class RegistryEndpointCaller<T> {
             || httpResponseException.getStatusCode()
                 == HttpStatusCodes.STATUS_CODE_METHOD_NOT_ALLOWED) {
           // The name or reference was invalid.
-          ErrorResponseTemplate errorResponse =
-              JsonTemplateMapper.readJson(
-                  httpResponseException.getContent(), ErrorResponseTemplate.class);
-          RegistryErrorExceptionBuilder registryErrorExceptionBuilder =
-              new RegistryErrorExceptionBuilder(
-                  registryEndpointProvider.getActionDescription(), httpResponseException);
-          for (ErrorEntryTemplate errorEntry : errorResponse.getErrors()) {
-            registryErrorExceptionBuilder.addReason(errorEntry);
-          }
-
-          throw registryErrorExceptionBuilder.build();
+          throw newRegistryErrorException(httpResponseException);
 
         } else if (httpResponseException.getStatusCode() == HttpStatusCodes.STATUS_CODE_FORBIDDEN) {
           throw new RegistryUnauthorizedException(
@@ -318,5 +308,30 @@ class RegistryEndpointCaller<T> {
       }
       throw ex;
     }
+  }
+
+  @VisibleForTesting
+  RegistryErrorException newRegistryErrorException(HttpResponseException httpResponseException) {
+    RegistryErrorExceptionBuilder registryErrorExceptionBuilder =
+        new RegistryErrorExceptionBuilder(
+            registryEndpointProvider.getActionDescription(), httpResponseException);
+
+    try {
+      ErrorResponseTemplate errorResponse =
+          JsonTemplateMapper.readJson(
+              httpResponseException.getContent(), ErrorResponseTemplate.class);
+      for (ErrorEntryTemplate errorEntry : errorResponse.getErrors()) {
+        registryErrorExceptionBuilder.addReason(errorEntry);
+      }
+    } catch (IOException ex) {
+      registryErrorExceptionBuilder.addReason(
+          "registry returned error code "
+              + httpResponseException.getStatusCode()
+              + "; possible causes include invalid or wrong reference. Actual error output follows:\n"
+              + httpResponseException.getContent()
+              + "\n");
+    }
+
+    return registryErrorExceptionBuilder.build();
   }
 }

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -515,6 +515,41 @@ public class RegistryEndpointCallerTest {
     Assert.assertFalse(RegistryEndpointCaller.isBrokenPipe(exception));
   }
 
+  @Test
+  public void testNewRegistryErrorException_jsonErrorOutput() {
+    HttpResponseException httpException = Mockito.mock(HttpResponseException.class);
+    Mockito.when(httpException.getContent())
+        .thenReturn(
+            "{\"errors\": [{\"code\": \"MANIFEST_UNKNOWN\", \"message\": \"manifest unknown\"}]}");
+
+    RegistryErrorException registryException =
+        secureEndpointCaller.newRegistryErrorException(httpException);
+    Assert.assertSame(httpException, registryException.getCause());
+    Assert.assertEquals(
+        "Tried to actionDescription but failed because: manifest unknown | If this is a bug, "
+            + "please file an issue at https://github.com/GoogleContainerTools/jib/issues/new",
+        registryException.getMessage());
+  }
+
+  @Test
+  public void testNewRegistryErrorException_nonJsonErrorOutput() {
+    HttpResponseException httpException = Mockito.mock(HttpResponseException.class);
+    // Registry returning non-structured error output
+    Mockito.when(httpException.getContent()).thenReturn(">>>> (404) page not found <<<<<");
+    Mockito.when(httpException.getStatusCode()).thenReturn(404);
+
+    RegistryErrorException registryException =
+        secureEndpointCaller.newRegistryErrorException(httpException);
+    Assert.assertSame(httpException, registryException.getCause());
+    Assert.assertEquals(
+        "Tried to actionDescription but failed because: registry returned error code 404; "
+            + "possible causes include invalid or wrong reference. Actual error output follows:\n"
+            + ">>>>> (404) page not found <<<<<\n"
+            + " | If this is a bug, please file an issue at "
+            + "https://github.com/GoogleContainerTools/jib/issues/new",
+        registryException.getMessage());
+  }
+
   /**
    * Verifies that a response with {@code httpStatusCode} throws {@link
    * RegistryUnauthorizedException}.

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/registry/RegistryEndpointCallerTest.java
@@ -535,7 +535,7 @@ public class RegistryEndpointCallerTest {
   public void testNewRegistryErrorException_nonJsonErrorOutput() {
     HttpResponseException httpException = Mockito.mock(HttpResponseException.class);
     // Registry returning non-structured error output
-    Mockito.when(httpException.getContent()).thenReturn(">>>> (404) page not found <<<<<");
+    Mockito.when(httpException.getContent()).thenReturn(">>>>> (404) page not found <<<<<");
     Mockito.when(httpException.getStatusCode()).thenReturn(404);
 
     RegistryErrorException registryException =

